core/domain/feature_flag_domain.py

```diff
@@ -0,0 +1,445 @@
+# coding: utf-8
+#
+# Copyright 2023 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Domain objects for feature flags."""
+
+from __future__ import annotations
+
+import datetime
+import enum
+import json
+import re
+
+from core import feconf
+from core import utils
+from core.constants import constants
+
+from typing import Final, List, Optional, TypedDict
+
+
+class ServerMode(enum.Enum):
+    """Enum for server modes."""
+
+    DEV = 'dev'
+    TEST = 'test'
+    PROD = 'prod'
+
+
+FeatureStages = ServerMode
+
+ALLOWED_FEATURE_STAGES: Final = [
+    FeatureStages.DEV.value,
+    FeatureStages.TEST.value,
+    FeatureStages.PROD.value
+]
+
+
+def get_server_mode() -> ServerMode:
+    """Returns the current server mode.
+
+    Returns:
+        ServerMode. The current server mode.
+    """
+    return (
+        ServerMode.DEV
+        if constants.DEV_MODE
+        else ServerMode.PROD
+        if feconf.ENV_IS_OPPIA_ORG_PRODUCTION_SERVER
+        else ServerMode.TEST
+    )
+
+
+class FeatureFlagDict(TypedDict):
+    """Dictionary representing FeatureFlag object."""
+
+    name: str
+    description: str
+    feature_stage: str
+    force_enable_for_all_users: bool
+    rollout_percentage: int
+    user_group_ids: List[str]
+    last_updated: Optional[str]
+
+
+class FeatureFlag:
+    """FeatureFlag domain object."""
+
+    def __init__(
+        self,
+        feature_flag_spec: FeatureFlagSpec,
+        feature_flag_value: FeatureFlagValue
+    ):
+        self._feature_flag_spec = feature_flag_spec
```

Comment:
I think the name should be top-level here.

Why do you need FeatureFlagSpec -- why can't description and feature stage also be top-level? Are there cases when the spec needs to be edited separately from the flag itself at runtime? (If so then that would justify the creation of a separate object, but if not then I'm not sure why you need it.)

Agreed with having a separate feature flag value though -- but here, also, the naming is off, the "value" of a feature flag is either True/False. I think you want feature_flag_config or similar.

Reply:
I have updated it.

The reason we require `FeatureFlagSpec` is because this is what we store in the registry variable. We do not want to store these value to the datastore hence we plan to store it in the registry variable.